File: time_test/satellite.library/python/short_float.py
# time_test/satellite.library/python/short_float.py -- the Python equivalent of
# programs/short_float.satl: f = f + 0.5, 200,000 turns.
# A satellite float is a decimal.Decimal (the steps of 0.5 are exact, but the semantics are
# satl's); Python has no decimal literal, so the 0.5 in the loop is made once before it, as satl
# makes its literal once when it reads the program. Otherwise line for line.
# The blank line printed first is the one satl prints after its line of dashes, so the answers
# compare byte for byte.

from decimal import ROUND_HALF_UP, Decimal, getcontext

# A SATELLITE FLOAT: 128 decimal places, rounded half away from zero where an operation does not
# end; 400 digits of working precision so a sum is exact before any rounding.
getcontext().prec = 400
getcontext().rounding = ROUND_HALF_UP
SHOWN = Decimal("1E-32")
HALF = Decimal("0.5")


def display(value):
    # satl shows a float rounded half away from zero to 32 places, trailing zeros removed.
    shown = format(value.quantize(SHOWN, rounding=ROUND_HALF_UP), "f")
    if "." in shown:
        shown = shown.rstrip("0").rstrip(".")
    print(shown)


def main():
    f = Decimal("0.5")
    counter = 0
    while counter < 200000:
        f = f + HALF
        counter = counter + 1
    display(f)
    return


print()
main()

File: time_test/satellite.library/cpp/short_float.cpp
// cpp/short_float.cpp -- the C++ equivalent of programs/short_float.satl: f = f + 0.5, 200,000 turns.
// Not line for line: it first prints the blank line satl prints after its dashes. f is a double,
// not satl's 128-place float: every value is a multiple of 0.5 under 2^53, so a double holds each
// one exactly. display() prints it the satl way: 32 places, trailing zeros dropped, one digit
// kept after the point (satl shows 2.0, 0.25, 100000.5). counter never passes 200,000.
//
// BUILD: clang++ -std=c++20 -O2 short_float.cpp -o short_float      (g++ works the same)

#include <cstdio>
#include <cstring>

namespace {

// satl's float display: rounded to 32 places, then the zeros at the end of them removed,
// leaving at least one digit after the point. %.32f is exact for these values (steps of 0.5).
void display(double f)
{
    char text[400];
    std::snprintf(text, sizeof text, "%.32f", f);
    char *point = std::strchr(text, '.');
    char *end = text + std::strlen(text);
    while (end - point > 2 && end[-1] == '0') --end;
    *end = '\0';
    std::printf("%s\n", text);
}

} // namespace

int main()
{
    std::printf("\n");

    double f = 0.5;
    long long counter = 0;
    while (counter < 200000) {
        f = f + 0.5;
        counter = counter + 1;
    }
    display(f);
    return 0;
}
